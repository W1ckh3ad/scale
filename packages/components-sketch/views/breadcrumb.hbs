{{#* inline "icon"}}m7.4 19.95c-.25-.35-.2-.8.1-1.05l8.8-6.9-8.75-6.9a.691.691 0 0 1 -.1-1.05c.25-.3.75-.4 1.05-.1l10.2 8.05-10.25 8.1c-.3.25-.8.2-1.05-.15z{{/inline}}

<div data-sketch-artboard="Breadcrumb" class="artboard">
    <h1 class="title--artboard">Breadcrumb</h1>
    
    <h2 class="title--artboard-section">Current Page</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:available</p>
            <scale-breadcrumb data-sketch-key="breadcrumb-1">
                <span is-current>Current Page</span>
            </scale-breadcrumb>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            —
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            –
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            –
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            –
        </div>
    </div>

    <h2 class="title--artboard-section">Previous Page</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:available</p>
            <scale-breadcrumb data-sketch-key="breadcrumb-1">
                <a href="#">Previous Page</a>
            </scale-breadcrumb>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-breadcrumb data-sketch-key="breadcrumb-2" data-sketch-state="hover">
                <a href="#">Previous Page</a>
            </scale-breadcrumb>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-breadcrumb data-sketch-key="breadcrumb-3" data-sketch-state="focus">
                <a href="#">Previous Page</a>
            </scale-breadcrumb>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-breadcrumb data-sketch-key="breadcrumb-4" data-sketch-state="active">
                <a href="#">Previous Page</a>
            </scale-breadcrumb>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            –
        </div>
    </div>

    <h2 class="title--artboard-section">Separator</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:available</p>
            <scale-icon path="{{>icon}}"></scale-icon>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            –
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            –
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            –
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            –
        </div>
    </div>
</div>

<script src="/symbol_names.js"></script>
<script>
    Array.from(document.querySelectorAll("scale-breadcrumb")).forEach(btn => {
        btn.dataset.sketchSymbol = `Breadcrumb / ${
                order(1, btn.firstElementChild.hasAttribute("is-current") ? 'Current Page' : 'Previous Page')
            }${
                btn.firstElementChild.hasAttribute("is-current") 
                ? ""
                : " / " + order(4, getStateName(btn))
            }`;
    });
    Array.from(document.querySelectorAll("scale-icon")).forEach(btn => {
        btn.dataset.sketchSymbol = `Breadcrumb / 03 Separator`;
    });
</script>