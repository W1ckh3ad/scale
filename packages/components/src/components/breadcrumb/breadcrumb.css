.breadcrumb {
  margin-left: calc(-1 * var(--scl-spacing-8));
}

ol {
  display: flex;
  list-style: none;
  margin-top: 0;
  padding-left: 0;
  margin-bottom: 0;
}

li {
  display: inline-flex;
  font-size: var(--scl-spacing-14, 0.875rem);
  align-items: center;
}

.separator {
  display: flex;
  color: var(--scl-color-grey-60);
}

.item,
.link {
  padding: var(--scl-spacing-4) var(--scl-spacing-8);
}

.link {
  text-decoration: none;
  color: var(--scl-color-grey-60);
  border-radius: var(--scl-radius-8);
}

.link:hover {
  color: var(--scl-color-primary-hover, #f90984);
}

.link:active {
  color: var(--scl-color-primary-active, #cb0068);
}

.link:focus {
  outline: none;
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}

.current {
  color: var(--scl-color-text-standard);
}
