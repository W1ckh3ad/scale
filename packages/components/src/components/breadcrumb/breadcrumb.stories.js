import { withKnobs, text } from '@storybook/addon-knobs';
import { h } from 'jsx-dom';

export default {
  title: 'Components/Breadcrumb',
  component: 'scale-breadcrumb',
  decorators: [withKnobs]
}

/* TODO show focus state  */

export const Playground = () => {
  return (
    <scale-breadcrumb separator={text('Separator')}>
      <a href="#1" id="one">First link</a>
      <a href="#2">Second link</a>
      <a href="#3">Third link</a>
    </scale-breadcrumb>
  )
}

export const WithIcon = () => {
  return (
    <scale-breadcrumb>
      <a href="#1">First link</a>
      <a href="#2">Second link</a>
      <a href="#3">Third link</a>
      <span slot="separator">
        <scale-icon
          size="18"
          path="M8.306 22.679a1.137 1.137 0 01-.061-1.481l.06-.07 7.741-8.126-7.74-8.13a1.137 1.137 0 01-.061-1.481l.06-.07c.387-.406 1-.427 1.411-.064l.066.064L19 13.001 9.782 22.68a1.01 1.01 0 01-1.476 0z"
        />
      </span>
    </scale-breadcrumb>
  )
}

export const CurrentWithNoLink = () => {
  return (
    <scale-breadcrumb>
      <a href="#1">First link</a>
      <a href="#2">Second link</a>
      <span>Third, just text</span>
    </scale-breadcrumb>
  )
}
