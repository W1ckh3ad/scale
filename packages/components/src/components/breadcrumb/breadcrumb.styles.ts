import { JssStyle } from 'jss';

export const styles: JssStyle = {
  breadcrumb: {
    marginLeft: ({ spacing }) => `-${spacing['2']}`,
    '& ol': {
      listStyle: 'none',
      paddingLeft: 0,
      marginTop: 0,
      marginBottom: 0,
      display: 'flex',
    },
    '& li': {
      display: 'inline-flex',
      alignItems: 'center',
    },
  },
  separator: {
    display: 'flex',
  },
  item: {
    padding: ({ spacing }) => `${spacing['1']} ${spacing['2']}`,
  },
  link: {
    color: ({ color }) => color.text,
    textDecoration: 'none',
    padding: ({ spacing }) => `${spacing['1']} ${spacing['2']}`,
    borderRadius: ({ radii }) => radii.medium,
    '&:hover': {
      color: ({ color }) => color.primary_hover,
    },
    '&:active': {
      color: ({ color }) => color.primary_active,
    },
    '&:focus': {
      boxShadow: ({ size, color }) =>
        `0 0 0 ${size.border_focus} ${color.focus}`,
      outline: 'none',
    },
  },
  current: {
    fontWeight: ({ type }) => type.weight_extrabold,
  },
};
