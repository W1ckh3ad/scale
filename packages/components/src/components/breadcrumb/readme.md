# scale-breadcrumb



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute   | Description                    | Type     | Default     |
| ----------- | ----------- | ------------------------------ | -------- | ----------- |
| `separator` | `separator` |                                | `string` | `'\\'`      |
| `styles`    | `styles`    | (optional) Injected jss styles | `any`    | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
