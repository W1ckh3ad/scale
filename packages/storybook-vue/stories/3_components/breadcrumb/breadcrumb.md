# Breadcrumb

Breadcrumbs help users find their current position in the information architecture of your product.

## Elements

![Image Name](/assets/3_components/breadcrumb/image-20200812110334888.png)

### Current Page (1)

The current page is highlighted at the right end of the breadcrumb.

### Parent Pages (2)

These are links that point to the respective page.

### Divider (3)

This element visually separates the individual pages.

## Code



