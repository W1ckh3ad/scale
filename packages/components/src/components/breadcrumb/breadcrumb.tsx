import { Component, h, State, Element, Prop, Host } from '@stencil/core';

import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './breadcrumb.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

/*
  @see https://www.w3.org/TR/wai-aria-practices/examples/breadcrumb/index.html
*/

@Component({
  tag: 'scale-breadcrumb',
  shadow: true,
})
export class Breadcrumb implements Base {
  @Element() hostElement: HTMLElement;

  @Prop() separator?: string;

  /** (optional) Injected jss styles */
  @Prop() styles?: any;
  /** decorator Jss stylesheet */
  @CssInJs('Breadcrumb', styles) stylesheet: StyleSheet;

  @State() linksArray = [];
  @State() separatorSlot: HTMLElement = null;

  componentWillLoad() {
    if (this.linksArray.length === 0) {
      this.setLinksArray();
    }
    this.separatorSlot = this.hostElement.querySelector('[slot="separator"]');
  }

  componentWillUpdate() {}
  componentDidUnload() {}

  setLinksArray() {
    this.linksArray = Array.from(this.hostElement.children).filter(
      element => element.slot === ''
    );
  }

  tmpDefaultIcon(size: number) {
    return (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width={size}
        height={size}
        viewBox="0 0 24 24"
        style={{ display: 'inline-flex' }}
      >
        <path
          d="M7.4 19.95c-.25-.35-.2-.8.1-1.05l8.8-6.9-8.75-6.9a.691.691 0 01-.1-1.05c.25-.3.75-.4 1.05-.1L18.7 12 8.45 20.1c-.3.25-.8.2-1.05-.15z"
          fill-rule="evenodd"
          fill="#6B6B6B"
        ></path>
      </svg>
    );
  }

  render() {
    const { classes } = this.stylesheet;
    const isLast = index => index === this.linksArray.length - 1;
    // Set aria-current="page" to the last item if it's a link
    const getCurrentAttr = index =>
      isLast(index) === true ? { 'aria-current': 'page' } : undefined;

    return (
      <Host>
        <style>{this.stylesheet.toString()}</style>
        <nav aria-label="Breadcrumb" class={this.getCssClassMap()}>
          <ol>
            {this.linksArray.map((element, index) => {
              const separator =
                this.separatorSlot != null ? (
                  <span
                    class={classes.separator}
                    innerHTML={this.separatorSlot.innerHTML}
                  />
                ) : (
                  <span class={classes.separator}>
                    {this.separator || (
                      <scale-icon-navigation-right size={16} color="#6B6B6B" />
                    )}
                  </span>
                );
              return (
                <li>
                  {element.href ? (
                    <a
                      href={element.href}
                      class={classNames(
                        isLast(index) && classes.current,
                        classes.link
                      )}
                      {...getCurrentAttr(index)}
                    >
                      {element.textContent}
                    </a>
                  ) : (
                    <span
                      class={classNames(
                        isLast(index) && classes.current,
                        classes.item
                      )}
                    >
                      {element.textContent}
                    </span>
                  )}
                  {!isLast(index) ? separator : null}
                </li>
              );
            })}
          </ol>
        </nav>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(classes.breadcrumb);
  }
}
