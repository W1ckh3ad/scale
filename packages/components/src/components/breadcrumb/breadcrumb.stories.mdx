import { Meta, Story, Description, Preview } from '@storybook/addon-docs/blocks';
import * as stories from './breadcrumb.stories.js';
import readme from './readme.md'

<Meta title="Components/Breadcrumb" component="scale-breadcrumb"/>

# Scale breadcrumb
â€¦

<Preview>
  <Story name="Playground">
    {stories.Playground()}
  </Story>
</Preview>

<Description>
  {readme}
</Description>
