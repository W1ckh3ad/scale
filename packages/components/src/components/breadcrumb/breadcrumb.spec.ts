/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { Breadcrumb } from './breadcrumb';
// import { styles } from './breadcrumb.styles';
// import jss from 'jss';

describe('Breadcrumb', () => {
  /* let element;
  let stylesheet;
  beforeEach(async () => {
    element = new Breadcrumb();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  }); */

  it('should match snapshot', async () => {
    let component: any;
    const page = await newSpecPage({
      components: [Breadcrumb],
      html: `<div></div>`,
    });
    component = page.doc.createElement('scale-breadcrumb');
    page.root.appendChild(component);
    expect(page.root).toMatchSnapshot();
  });
});
