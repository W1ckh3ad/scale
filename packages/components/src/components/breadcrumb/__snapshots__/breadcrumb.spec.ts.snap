// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb should match snapshot 1`] = `
<scale-breadcrumb>
  <mock:shadow-root>
    <nav aria-label="Breadcrumb" class="breadcrumb">
      <ol></ol>
    </nav>
  </mock:shadow-root>
</scale-breadcrumb>
`;
